### DISCUSSION TITLE
How to query on JSON field type with an object that contains at least the properties stored in the database

### DISCUSSION THREAD

**Discussion by mhmdsalahsebai** [prisma/prisma#24004] [Q&A]
### Question

I have a Prisma model defining events stored in a PostgreSQL database. One of the fields is additionalFields, which is a JSON object. How can I construct a Prisma query to retrieve events, matching even if the additionalFields contain additional properties beyond those specified in the query? Here's my model:
```
model Event {
  id               Int            @id @default(autoincrement())
  eventType        String
  additionalFields Json
}

```

Suppose I have an event in the database with the following data:
```
{
id: 1,
eventType: "type1",
additionalFields: {
  "name": "This is a name",
  "amount": 123
} 
```
I want to construct a Prisma query that can match this event even if the additionalFields contain extra properties. For instance, the query below should match the event above:

```
const dataToQuery {
      eventType: "type1",
      additionalFields: {
       "name": "This is a name",
       "amount": 123,
       "foo": "bar",
       "whatever": "value"
      }
}
```

but this should not match,since there is a missing property:
```
const dataToQuery {
      eventType: "type1",
      additionalFields: {
       "name": "This is a name",
      }
}
```

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @mhmdsalahsebai 👋 

Prisma Client does not support partial matching or ignoring additional properties in the JSON field. The JSON field in the query must exactly match the JSON field in the database.

For Example:
```js
const getEvents = await prisma.event.findMany({  
 where: {  
  additionalFields: {  
   path: ['name'],  
   equals: 'This is a name',  
  },  
 },  
})
```
The query returns all events where the `name` property in additionalFields is `This is a name`.
However, this does not meet your requirement of matching events even if the additionalFields contain extra properties. Prisma Client does not currently support this feature. 

However, you can open a [feature request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) and document your usecase so our engineers can take a look.

**Comment by mhmdsalahsebai** [comment]
Hi @ludralph,

Thanks for your response.

Do you know if this  is possible with raw SQL? I searched a lot and I didn't find anything thanks.

**Reply by ludralph** [reply]
Hi @mhmdsalahsebai 

Yes, this is possible with raw SQL in PostgreSQL. PostgreSQL provides powerful JSON operators that allow for more flexible querying of JSON data, including the ability to match objects that contain additional properties beyond those specified in the query.

The key operator for this scenario is the `@>` (contains) operator. This operator checks if the left JSON value contains the right JSON value as a subset. This is exactly what you're looking for - it will match even if there are additional properties in the database that aren't specified in your query.

For example:
```sql
SELECT * FROM "Event"
WHERE "eventType" = 'type1'
  AND "additionalFields" @> '{"name": "This is a name", "amount": 123}'::jsonb;
```
This query will match any event where:

- The eventType is 'type1'
- The additionalFields JSON contains at least the specified properties ("name" and "amount") with the exact values given.

It will match even if additionalFields contains additional properties beyond those specified. To use this with Prisma, you can execute a raw query like this:
```js
const result = await prisma.$queryRaw`
  SELECT * FROM "Event"
  WHERE "eventType" = 'type1'
    AND "additionalFields" @> '{"name": "This is a name", "amount": 123}'::jsonb;
`;
```

**Reply by JPustkuchen** [reply]
See https://hashrocket.com/blog/posts/dealing-with-nested-json-objects-in-postgresql for a nice article on these JSON operators,

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;

**Comment by JPustkuchen** [comment]
Also see https://github.com/prisma/prisma/issues/17852 maybe?
